### PR TITLE
Get google sheets data

### LINKED ISSUE #209: getKoboData only for forms in GoogleSheet
## Background, context, and business value
This builds on issue #206 

Right now WCS has jobs like this [getKoboData](https://github.com/OpenFn/ConSoSci/blob/master/bns/getKoboData.js#L14-L35) job that runs every 3 hours to fetch Kobo data submissions from the API. Every time Diane wants to sync data for a Kobo form, she has to manually edit the job code to edit the list of form ids. 

Now that we have [this google sheet](https://docs.google.com/spreadsheets/d/1s7K3kxzm5AlpwiALattyc7D9_aIyqWmo2ubcQIUlqlY/edit#gid=0) which will later contain ALL of these same form ids has L14-35... so let's remove these lines and instead send a GET to the GoogleSheet to dynamically get the list of Kobo forms to query. 

## The specific request, in as few words as possible
1. Replace [L14-L35](https://github.com/OpenFn/ConSoSci/blob/master/bns/getKoboData.js#L14-L35) with a step that gets the list of rows from [this google sheet](https://docs.google.com/spreadsheets/d/1s7K3kxzm5AlpwiALattyc7D9_aIyqWmo2ubcQIUlqlY/edit#gid=0)
2. Filter the rows to only return the form rows where `automate_sync = TRUE` and `tag = 'bns_survey' || 'nrgt_current'`. 
Only if these^ criteria are true, should we then move to the next step of the job where we [get data for these form ids](https://github.com/OpenFn/ConSoSci/blob/master/bns/getKoboData.js#L51-L60). 

### GoogleSheets 
- configuration: For now use your Google user - you have access to [this sheet](https://docs.google.com/spreadsheets/d/1s7K3kxzm5AlpwiALattyc7D9_aIyqWmo2ubcQIUlqlY/edit#gid=0)
- data: { new kobo forms detected}

####Kobo 
- configuration: see LP `OpenFn Kobo Toolbox (for WCS data access!)`
- data: {}

## adaptor
http - let's update the job to work on latest version of http 

## expression.js
create a new branch off of `master`
https://github.com/OpenFn/ConSoSci/blob/master/bns/getKoboData.js#L51-L60

## output.json
GET request should be sent to get Kobo data form all form rows in the Google Sheet that match the requested criteria

### PR DESCRIPTION
**Pull request by hunterachieng**
## Summary

Write jobs to allow a user to fetch data from google sheets and then use them to get more form details from Kobo

## Implementation Details

Currently a manual and tedious process is being used to map and write the current google Sheets data. This implementation will focus on writing a job to automatically fetch google sheets data.

## Issues

Fixes #209

### QA Check

**1. Get google sheets data**
>adaptor: googlesheets@2.3.0
```json
{
  "accessToken": "Use 'gcloud auth print-access-token'" 
}

```
Read the following document as a quick start guide to creating your own access token
[Goole OAuth](https://blog.postman.com/how-to-access-google-apis-using-oauth-in-postman/)

### CODE REVIEW

File: bns/getKoboData.js [hunk 1]
```diff
@@ -1,45 +1,45 @@
...
+    const filteredSheetsData = values.filter(item => {
+      if (
+        item.includes('TRUE') &&
+        item.includes('bns_survey', 'nrgt_current')
+      ) {
+        return item;
+      }
+    });
+
+    state.sheetsData = filteredSheetsData.map(item => {
```
Comment (mtuchi) [inline bns/getKoboData.js:19]: Hiya @hunterachieng i think we can improve this logic here 

```
const mapHeaderToValue = (value) => {
  headers.reduce((obj, header) => {
    obj[header] = value[headers.indexOf(header)];
    return obj;
  }, {});
};

state.sheetsData = values
  .filter(
    (item) =>
      item.includes("TRUE") && item.includes("bns_survey", "nrgt_current")
  )
  .map((item) => mapHeaderToValue(item));

```
I am not sure if `mapHeaderToValue()` is the right function name but the idea is to decouple the mapping of key value pair for each column.

File: bns/getKoboData.js [hunk 2]
```diff
@@ -1,45 +1,45 @@
...
+  'sheetsList of Kobo forms!A:L',
+  state => {
+    const [headers, ...values] = state.data.values;
+
+    const filteredSheetsData = values.filter(item => {
+      if (
+        item.includes('TRUE') &&
+        item.includes('bns_survey', 'nrgt_current')
+      ) {
+        return item;
```
Comment (mtuchi) [inline bns/getKoboData.js:15]: You don't have to say `return item`. The callback function that `array.filter(cbFn)` takes, should return a true value to keep the element in the resulting array. [See MDN  docs on Array.filter() to learn more](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)

File: bns/getKoboData.js [hunk 3]
```diff
@@ -1,45 +1,45 @@
...
+        obj[header] = item[headers.indexOf(header)];
+        return obj;
+      }, {});
+    });
+
+    return state;
+  }
+);
 
 alterState(state => {
```
Comment (mtuchi) [inline bns/getKoboData.js:30]: Please use `fn()` instead of ~alterState()~

**Review by mtuchi — CHANGES_REQUESTED** [review]
@hunterachieng i left some comments for you, also watch on the slack thread on implementation regarding refactoring the nested `each` operations. No action plan yet so don't implement any of the suggestion in that thread unless @aleksa-krolls gives you a go ahead

**Review by mtuchi — APPROVED** [review]
@hunterachieng this looks good

**Review by aleksa-krolls — APPROVED** [review]

**Pull request merged**